core/src/main/java/org/apache/calcite/sql/SqlJsonTypeNameSpec.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to you under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.calcite.sql;
+
+import org.apache.calcite.rel.type.RelDataType;
+import org.apache.calcite.sql.parser.SqlParserPos;
+import org.apache.calcite.sql.validate.SqlValidator;
+import org.apache.calcite.util.Litmus;
+
+/**
+ * A sql type name specification of the JSON type.
+ */
+public class SqlJsonTypeNameSpec extends SqlTypeNameSpec {
+
+  private final Integer maxLength;
+  private final Integer inlineLength;
+  private final CharacterSet characterSet;
+  private final StorageFormat storageFormat;
+
+  public SqlJsonTypeNameSpec(SqlParserPos pos,
+      Integer maxLength,
+      Integer inlineLength,
+      CharacterSet characterSet,
+      StorageFormat storageFormat) {
+    super(new SqlIdentifier("JSON", pos), pos);
+    this.maxLength = maxLength;
+    this.inlineLength = inlineLength;
+    this.characterSet = characterSet;
+    this.storageFormat = storageFormat;
+
+    // These two fields are mutally exclusive.
+    assert characterSet == null || storageFormat == null;
+
+    if (characterSet != null) {
+      assert characterSet == CharacterSet.LATIN
```

Comment:
Per feedback we've gotten from the rest of the team on previous asserts, they would prefer we just capture this requirement in comments rather than have the asserts.